Smarter astgen handling

max-leuthaeuser opened the pull request:
This improves the astgen binary handling.
The following rules apply:
1) If astgen is found on the system it is used without downloading binaries at all iff the version installed matches or is higher.
2) sbt stage downloads all binaries so the distribution always works on all platforms like before. This does not happen iff rule 1 triggers.
3) compile only downloads the astgen binary for the system it's invoked on.

This is for: https://github.com/joernio/joern/issues/2205

iandinwoodie commented:
@max-leuthaeuser is there is any chance that item 1 could be expanded to:

> 1. If astgen is found on the system it is used without downloading binaries at all iff **a compatible version is found**.

where a compatible version satisfies both:

1. `found.major == required.major`
2. `found.minor >= required.minor`

If we require an exact match in version, then there is going to be an issue when it comes time to bump the version of the `joern` and `astgen` formulas. For example:

* If the `astgen` formula is bumped first, then `joern` would find a newer `astgen` in the PATH and will resort to fetching binaries until its own formula is bumped.
* If the `joern` formula is bumped first, then the `astgen` built by Homebrew will be older than required and the build system will again resort to fetching binaries until the `astgen` formula is bumped.

I do not believe the Homebrew maintainers allow for two formulas to be bumped in the same PR. If we instead check for a compatible version (as indicated by SemVer), then we would be able to bump the `astgen` formula and then the `joern` formula without ever resorting to fetching binaries.

The edge case here is when `found.major != required.major`. A possible solution for this is to add a means (e.g., command line option or environment variable) for us to fail the build if a compatible `astgen` version is not found. This means that the `joern` formula will fail until it is bumped whenever `astgen` has incompatible API changes, but I think that sounds reasonable.

***

An alternative solution is to implement item 1 as-is, but include the means to fail the build when an exact match is not found. I could then pin an exact `astgen` version in the `joern` formula to keep the two formula in sync. This way `joern` built with Homebrew either builds with a Homebrew built `astgen` or fails. The only drawback to this approach is that if you have `astgen 3.2.0` installed, but `joern` has `astgen 3.1.0` pinned, then you’re going to end up with both `astgen 3.1.0` and `astgen 3.2.0` installed even though `astgen 3.2.0` should have been sufficient (according to SemVer).

max-leuthaeuser commented:
@iandinwoodie Thanks for spotting this issue! You are right.

Astgen is using SemVer versioning but its sufficient to check for a version installed >= the version requested by joern. We do not need to distinguish between major/minor versions as I will never release a new astgen without having that specific new version tested within joern and updated there as well.

I updated this PR to do that comparison.

mpollmeier commented on joern-cli/frontends/jssrc2cpg/src/main/scala/io/joern/jssrc2cpg/utils/AstGenRunner.scala:
```diff
@@ -6,26 +6,83 @@ import io.joern.jssrc2cpg.preprocessing.EjsPreprocessor
     skippedFiles: List[(String, String)] = List.empty
   )
+
+  private val ExecutableName = Environment.OperatingSystem match {
```
style, here and in other places: the value name should be lowercase, i.e. `val executableName`. Only constants are uppercase by convention, e.g. `val MagicNumber = 42`

max-leuthaeuser replied:
Right, will change that 👍
